src/flow_mc.py

```diff
@@ -0,0 +1,324 @@
+        """
+        self.__dict__.update(flow_dict)
+        # Set attributes of FlowSimGeneric (env, fields, yaml_template, sim_param_range, geo_file).
+        self.mesh_step = mesh_step
+        self._make_mesh()
+        self._substitute_yaml()
+
+
+    def _make_mesh(self):
+        """
+        Make the mesh, mesh_file: <geo_base>_step.msh.
+        Make substituted yaml: <yaml_base>_step.yaml,
+        using common fields_step.msh file for generated fields.
+        :return:
+        """
+        subprocess.call(self.env.gmsh, '-2', '-clscale', self.mesh_step, self.geo_file)
+        self.mesh_file = os.path.splitext(self.geo_file)[0] + '.msh'
+        mesh = gmsh_io.GmshIO(self.mesh_file)
+        n_ele = len(mesh.elements)
+        self.points = np.zeros(n_ele, 3)
+        self.ele_ids = np.zeros(n_ele)
+        i = 0
+        for id, el in mesh.elements.items():
+            type, tags, i_nodes = el
+            self.points[i] = np.average(np.array([mesh.nodes[i_node] for i_node in i_nodes]), axis=0)
+            self.ele_ids[i] = id
+            i += 1
+
+
+    def _substitute_yaml(self):
+        """
+        Create substituted YAML file from the tamplate.
+        :return:
+        """
+        self.yaml_file = os.path.splitext(self.yaml_template)[0] + '.inst.yaml'
+        self.fields_file = os.path.splitext(self.yaml_template)[0] + '.data.msh'
+        param_dict = {}
+        field_tmpl = "!FieldElementwise {gmsh_file: ${INPUT}/%s, field_name: %s}"
+        for field in self.fields.names():
+            param_dict[field] = field_tmpl%(self.fields_file, field)
+        param_dict['mesh_file'] = self.mesh_file
+        sub_file_params(self.yaml_template, self.yaml_file, param_dict)
+
+
+    def set_coarse_sim(self, coarse_sim):
+        """
+        Set coarse simulation ot the fine simulation so that the fine can generate the
+        correlated input data sample for both.
+
+        Here in particular set_points to the field generator
+        :param coarse_sim
+        """
+        coarse_centers = coarse_sim.points
+        both_centers = np.concatenate((self.points, coarse_centers), axis=0)
+        self.n_fine_elements = self.points.shape[0]
+        self.fields.set_points(both_centers)
+
+
+    # Needed by Level
+    def random_array(self):
+        """
+        Generate random field, both fine and coarse part.
+        Store them separeted.
+        :return:
+        """
+        assert self._is_fine_sim
+        fields_sample = self.fields.sample()
+        self._input_sample = {name: values[:self.n_fine_elements] for name, values in fields_sample }
+        self._coarse_sample = {name: values[self.n_fine_elements:] for name, values in fields_sample }
+
+    def get_random_array(self):
+        """
+        Return coarse part of generated field.
+        :return:
+        """
+        return self._coarse_sample
+
+    def cycle(self, sample_tag):
```

Comment:
sample_tag má být unikátní v rámci jednoho spuštění mlmc resp. flow nebo je možné, že se to bude celé spouštět nekolikrát (pro stejný work_dir_base)? V takovém připadě by bylo potřeba při spouštění cycle (na úrovni levelu) znát work_dir_base a nacházet unikatní ID.

Reply:
Staci unikatni v ramci jednoho spusteni.